### PR TITLE
Introduce internal types for sql blobs

### PR DESCRIPTION
**Pull request by andrewjdawson2016**
This is step 1 of the following 

**1. Introduce these types**
2. Add mappers which will convert to and from thrift types 
3. Add mappers which will convert to and from proto types 
4. Update the encoder interface methods to accept these internal types - the encoder will then use the mapper written in step 2 to convert to thrift blob before encoding. 
5. Update the decoder interface methods to return these internal types - the decoder will then use the mapper written in step 2 to convert from thrift blob to internal before returning. 
6. Implement encoder and decoder for proto just like we did for thrift

### CODE REVIEW

File: common/persistence/serialization/interfaces.go [hunk 1]
```diff
@@ -23,13 +23,281 @@
...
+		ParentDomainID                     *string
+		ParentWorkflowID                   *string
+		ParentRunID                        *string
+		InitiatedID                        *int64
+		CompletionEventBatchID             *int64
+		CompletionEvent                    []byte
+		CompletionEventEncoding            *string
+		TaskList                           *string
+		WorkflowTypeName                   *string
+		WorkflowTimeoutSeconds             *time.Duration
```
Comment (vytautas-karpavicius) [inline common/persistence/serialization/interfaces.go:110]: As this is `time.Duration`, rename it to `WorkflowTimeout`

File: common/persistence/serialization/interfaces.go [hunk 2]
```diff
@@ -23,13 +23,281 @@
...
+		DecisionTaskTimeoutSeconds         *time.Duration
```
Comment (vytautas-karpavicius) [inline common/persistence/serialization/interfaces.go:111]: Same here, rename it to `DecisionTaskTimeout`

File: common/persistence/serialization/interfaces.go [hunk 3]
```diff
@@ -23,13 +23,281 @@
...
+		DecisionScheduledTimestamp         *time.Time
+		CancelRequested                    *bool
+		DecisionOriginalScheduledTimestamp *time.Time
+		CreateRequestID                    *string
+		DecisionRequestID                  *string
+		CancelRequestID                    *string
+		StickyTaskList                     *string
+		StickyScheduleToStartTimeout       *time.Duration
+		RetryAttempt                       *int64
+		RetryInitialIntervalSeconds        *time.Duration
```
Comment (vytautas-karpavicius) [inline common/persistence/serialization/interfaces.go:137]: Same here, drop `Seconds`

File: common/persistence/serialization/interfaces.go [hunk 4]
```diff
@@ -23,13 +23,281 @@
...
+		RetryMaximumIntervalSeconds        *time.Duration
```
Comment (vytautas-karpavicius) [inline common/persistence/serialization/interfaces.go:138]: Same here, drop `Seconds`

File: common/persistence/serialization/interfaces.go [hunk 5]
```diff
@@ -23,13 +23,281 @@
...
+		RetryMaximumAttempts               *int32
+		RetryExpirationSeconds             *time.Duration
```
Comment (vytautas-karpavicius) [inline common/persistence/serialization/interfaces.go:140]: Same here, drop `Seconds`

File: common/persistence/serialization/interfaces.go [hunk 6]
```diff
@@ -23,13 +23,281 @@
...
+		RetryBackoffCoefficient            *float64
+		RetryExpirationTime                *time.Time
```
Comment (vytautas-karpavicius) [inline common/persistence/serialization/interfaces.go:142]: Just a suggestion, some `time.Time` fields have `Time` suffix while others uses `Timestamp` suffix. Maybe we should unify this?

File: common/persistence/serialization/interfaces.go [hunk 7]
```diff
@@ -23,13 +23,281 @@
...
+		ScheduledEvent                []byte
+		ScheduledEventEncoding        *string
+		ScheduledTime                 *time.Time
+		StartedID                     *int64
+		StartedEvent                  []byte
+		StartedEventEncoding          *string
+		StartedTime                   *time.Time
+		ActivityID                    *string
+		RequestID                     *string
+		ScheduleToStartTimeoutSeconds *time.Duration
```
Comment (vytautas-karpavicius) [inline common/persistence/serialization/interfaces.go:174]: For this and following lines below, drop `Seconds` suffix.

File: common/persistence/serialization/interfaces.go [hunk 8]
```diff
@@ -23,13 +23,281 @@
...
+		StartToCloseTimeoutSeconds    *time.Duration
+		HeartbeatTimeoutSeconds       *time.Duration
+		CancelRequested               *bool
+		CancelRequestID               *int64
+		TimerTaskStatus               *int32
+		Attempt                       *int32
+		TaskList                      *string
+		StartedIdentity               *string
+		HasRetryPolicy                *bool
+		RetryInitialIntervalSeconds   *time.Duration
```
Comment (vytautas-karpavicius) [inline common/persistence/serialization/interfaces.go:185]: Same here

File: common/persistence/serialization/interfaces.go [hunk 9]
```diff
@@ -23,13 +23,281 @@
...
 )
 
+type (
+	// ShardInfo blob in a serialization agnostic format
+	ShardInfo struct {
+		StolenSinceRenew                      *int32
+		UpdatedAt                             *time.Time
+		ReplicationAckLevel                   *int64
+		TransferAckLevel                      *int64
+		TimerAckLevel                         *time.Time
```
Comment (vytautas-karpavicius) [inline common/persistence/serialization/interfaces.go:42]: I think this should be `int64`

**Review by vytautas-karpavicius — COMMENTED** [review]
As all those are newly defines types which are serialization agnostic, maybe we should start dropping pointer for fields which are always present?

File: common/persistence/serialization/interfaces.go (hunk 9 above)
Comment (andrewjdawson2016) [reply common/persistence/serialization/interfaces.go:42]: In persistenceInterface.go this has been converted to time.Time.

File: common/persistence/serialization/interfaces.go [hunk 10]
```diff
@@ -23,13 +23,281 @@
...
+		DomainNotificationVersion             *int64
+		ClusterTransferAckLevel               map[string]int64
+		ClusterTimerAckLevel                  map[string]time.Time
```
[the remark this answers is not in the document]
Comment (andrewjdawson2016) [reply common/persistence/serialization/interfaces.go:45]: Same here.

File: common/persistence/serialization/interfaces.go [hunk 11]
```diff
@@ -23,13 +23,281 @@
...
+		TimerProcessingQueueStatesEncoding    *string
+	}
+
+	// DomainInfo blob in a serialization agnostic format
+	DomainInfo struct {
+		Name                        *string
+		Description                 *string
+		Owner                       *string
+		Status                      *int32
+		RetentionDays               *time.Duration
```
[the remark this answers is not in the document]
Comment (andrewjdawson2016) [reply common/persistence/serialization/interfaces.go:63]: Agreed

**Review by vytautas-karpavicius — APPROVED** [review]

**Pull request merged**